Feature/cart page checkout form

Ahmed-Fahmy-Dev opened the pull request:
-JSX
-Styling
-Tests

please check my commit messages for more details on changes and features added, all messages are semantic and detailed.

Ahmed-Fahmy-Dev committed:
feat: checkout form template model
- validation schema file created
- initial values file created
- form model created

Ahmed-Fahmy-Dev committed:
feat: checkout form dependencies installed
- formik-mui package installed
- formik -mui date pickers installed

Ahmed-Fahmy-Dev committed:
feat: checkout form pngs added
- visa and mastercard png added to assets

Ahmed-Fahmy-Dev committed:
feat: checkout form components exported
- component index.js file updated with new imports exports

Ahmed-Fahmy-Dev committed:
feat: checkout form added to checkout page

Ahmed-Fahmy-Dev committed:
feat: checkout form components created
- checkout form parent component
- userDetails, ShippingAddress, Payment child forms created
- Payment success modal component created
- Checkout summary component created

Ahmed-Fahmy-Dev committed:
feat: checkout form tests
- snapshot tests written for all components

Ahmed-Fahmy-Dev committed:
feat: checkout page snapshot test written

Ahmed-Fahmy-Dev committed:
Merge branch 'develop' into feature/cartPage-checkout-form

Ahmed-Fahmy-Dev committed:
Merge branch 'develop' into feature/cartPage-checkout-form

Ahmed-Fahmy-Dev committed:
Merge branch 'develop' into feature/cartPage-checkout-form

HIMDanny commented on client/src/components/CheckoutForm/CheckoutForm.js:
```diff
@@ -0,0 +1,66 @@
+import { Formik, Form } from 'formik';
+import { Stepper, Step, StepLabel, Button } from '@mui/material';
+import { UserDetailsForm, ShippingAddressForm, PaymentForm, PaymentSuccess, CheckoutSummary } from '..';
+import checkoutFormModel from '../../utils/checkoutFormModels/checkoutFormModel';
+import CheckoutFormInitialValues from '../../utils/checkoutFormModels/CheckoutFormInitialValues';
+import CheckoutFormvalidationSchema from '../../utils/checkoutFormModels/CheckoutFormvalidationSchema';
```
what about creating index in utils folder and import everything from it?

Ahmed-Fahmy-Dev replied:
Will be fixed in next file structure roll out

HIMDanny commented on client/src/components/CheckoutForm/CheckoutForm.js:
```diff
@@ -0,0 +1,66 @@
+    setActiveStep(activeStep + 1);
+  };
+
+  const GoTOPrevStep = () => {
```
```suggestion
  const GoToPrevStep = () => {
```

Ahmed-Fahmy-Dev replied:
spelling mistake fixed

HIMDanny commented on client/src/components/CheckoutForm/CheckoutForm.js:
```diff
@@ -0,0 +1,66 @@
+  const currentValidationSchema = CheckoutFormvalidationSchema[activeStep];
+
+  const GoToNextStep = () => {
+    setActiveStep(activeStep + 1);
```
suggest using previous value of active step:
```suggestion
    setActiveStep(prev => prev+ 1);
```

Ahmed-Fahmy-Dev replied:
call back used

HIMDanny commented on client/src/components/CheckoutForm/CheckoutForm.js:
```diff
@@ -0,0 +1,66 @@
+  };
+
+  const GoTOPrevStep = () => {
+    setActiveStep(activeStep - 1);
```
suggest using previous value of active step:
```suggestion
    setActiveStep(prev => prev - 1);
```

Ahmed-Fahmy-Dev replied:
call back used

HIMDanny commented on client/src/components/CheckoutForm/CheckoutForm.js:
```diff
@@ -0,0 +1,66 @@
+        >
+          {({ isSubmitting }) => (
+            <Form>
+              {activeStep === 0 && <UserDetailsForm checkoutFormModel={checkoutFormModel} />}
+              {activeStep === 1 && <ShippingAddressForm checkoutFormModel={checkoutFormModel} />}
+              {activeStep === 2 && <PaymentForm checkoutFormModel={checkoutFormModel} />}
```
you just passing this object to components. How about instead import it inside components?

Ahmed-Fahmy-Dev replied:
yup. i had it in my future hotfix list since the model itself is not needed in the parent form.  already fixed.

HIMDanny commented on client/src/utils/checkoutFormModels/CheckoutFormvalidationSchema.js:
```diff
@@ -0,0 +1,46 @@
+  cvv,
+} = checkoutFormModel;
+
+// const visaRegEx = /^(?:4[0-9]{12}(?:[0-9]{3})?)$/;
```
https://stackoverflow.com/questions/12011792/regular-expression-matching-a-3-or-4-digit-cvv-of-a-credit-card

Ahmed-Fahmy-Dev replied:
Yup matcher changed to string,  .matches regexp added for CVV

HIMDanny commented on client/src/utils/checkoutFormModels/checkoutFormModel.js:
```diff
@@ -0,0 +1,78 @@
+export default {
+  firstName: {
+    name: 'firstName',
+    label: 'First Name*',
```
you can get rid of `*` and use `required` prop in `Field`

Ahmed-Fahmy-Dev replied:
agree. done

HIMDanny commented on client/src/utils/checkoutFormModels/checkoutFormModel.js:
```diff
@@ -0,0 +1,78 @@
+  addressOne: {
+    name: 'addressOne',
+    label: 'Address line 1*',
+    requiredErrorMsg: 'Last name is required',
```
```suggestion
    requiredErrorMsg: 'Address is required',
```

Ahmed-Fahmy-Dev replied:
spelling mistake fixed

HIMDanny commented on client/src/components/CheckoutForm/CheckoutForm.test.js:
```diff
@@ -0,0 +1,18 @@
+import { render } from '@testing-library/react';
+import CheckoutForm from './CheckoutForm';
+
+jest.mock('@mui/material/Stepper', () => ({ children }) => <div>{children}</div>);
+jest.mock('@mui/material/Step', () => ({ children }) => <ul>{children}</ul>);
+jest.mock('@mui/material/StepLabel', () => ({ children }) => <li>{children}</li>);
+jest.mock('@mui/material/Button', () => ({ children }) => <button>{children}</button>);
+jest.mock('../../components/CheckoutSummary/CheckoutSummary', () => () => <div>CHECKOUT SUMMARY</div>);
+jest.mock('../../components/PaymentSuccess/PaymentSuccess', () => () => <div>PAYMENT SUCCESS</div>);
+jest.mock('../../components/UserDetailsForm/UserDetailsForm', () => () => <div>USER DETAILS FORM</div>);
+
```
mocked `ShippingAddressForm` and `PaymentForm` are missing

Ahmed-Fahmy-Dev replied:
they werent added on purpose since in this snapshot they dont render at all, activeStep is alwats 0 which in return renders only the userDetails form.

AlinaPalyvoda commented on client/src/components/UserDetailsForm/UserDetailsForm.js:
```diff
@@ -0,0 +1,36 @@
+import { TextField } from 'formik-mui';
+import { Field } from 'formik';
+
+const UserDetailsForm = ({ checkoutFormModel }) => {
+  const { firstName, lastName, email, phone } = checkoutFormModel;
+
+  return (
+    <>
+      <Typography variant="h3" align="center" sx={{ mb: '20px' }}>
+        User Details
+      </Typography>
+
+      <Grid container spacing={2}>
+        <Grid item xs={12} laptop={6}>
+          <Field component={TextField} name={firstName.name} label={firstName.label} fullWidth />
+        </Grid>
+
+        <Grid item xs={12} laptop={6}>
+          <Field component={TextField} name={lastName.name} label={lastName.label} fullWidth />
+        </Grid>
+
+        <Grid item xs={12} laptop={6}>
+          <Field component={TextField} name={email.name} label={email.label} fullWidth />
+        </Grid>
+
+        <Grid item xs={12} laptop={6}>
+          <Field component={TextField} name={phone.name} label={phone.label} fullWidth />
+        </Grid>
+      </Grid>
+    </>
```
maybe it will be more optimal? what do you think?

```suggestion
const UserDetailsForm = ({ checkoutFormModel }) => {
  const { firstName, lastName, email, phone } = checkoutFormModel;
  const data = [firstName, lastName, email, phone];
  return (
    <>
      <Typography variant="h3" align="center" sx={{ mb: '20px' }}>
        User Details
      </Typography>

      <Grid container spacing={2}>
          {data.map((el) => (
          <Grid key={el.name} item xs={12} laptop={6}>
            <Field component={TextField} name={el.name} label={el.label} fullWidth />
          </Grid>
        ))}
    </>
```

AlinaPalyvoda replied:
If you agree to use the above method, please apply it in ShippingAddressForm and PaymentForm

Ahmed-Fahmy-Dev replied:
dont agree, since these are gird items and sizing is controlled individually by props for different breakpoints, for better readability and control over each item, better to be represented individually not through a map method.

AlinaPalyvoda commented on client/src/components/CheckoutForm/CheckoutForm.js:
```diff
@@ -0,0 +1,66 @@
+import { UserDetailsForm, ShippingAddressForm, PaymentForm, PaymentSuccess, CheckoutSummary } from '..';
+import checkoutFormModel from '../../utils/checkoutFormModels/checkoutFormModel';
+import CheckoutFormInitialValues from '../../utils/checkoutFormModels/CheckoutFormInitialValues';
+import CheckoutFormvalidationSchema from '../../utils/checkoutFormModels/CheckoutFormvalidationSchema';
```
```suggestion
import CheckoutFormValidationSchema from '../../utils/checkoutFormModels/CheckoutFormValidationSchema';
```

Ahmed-Fahmy-Dev replied:
filename is validationSchema with a lowercase v,   this is a temporary naming scheme since it will be renamed after being moved to the file structure.

AlinaPalyvoda commented:
Text overlay occurs when you try to fill out a form with your previously saved data
![зображення_viber_2022-11-15_04-52-12-917](https://user-images.githubusercontent.com/107770819/201815887-02148885-d649-45be-8e79-988d46bbcc3f.jpg)

AlinaPalyvoda commented:
After filling the form, the fields in the next step are disabled
![image](https://user-images.githubusercontent.com/107770819/201816246-3244ebec-8cf3-4424-bd88-fafb0f404dd6.png)

AlinaPalyvoda commented:
I think this happens because after the first step, the formSubmitHandler function writes empty strings to value for the fields of the next steps
![image](https://user-images.githubusercontent.com/107770819/201816972-501ab84a-787e-43c2-ae22-d17ce9601fc4.png)

AlinaPalyvoda commented:
Also, the buttons are disabled when you return a step back.
![image](https://user-images.githubusercontent.com/107770819/201818337-412260d7-6de9-491c-8825-93abbbb0903f.png)

AlinaPalyvoda commented:
It would be nice to add some more validation. For example, that the city is a string, and the code is a number. I think this way the form will have a more finished look.

AlinaPalyvoda commented:
I suggest making the placeholder dimmer, like grey. What do you think?
![image](https://user-images.githubusercontent.com/107770819/201819552-0115a3aa-0953-4f3a-8212-be9010b6193c.png)

Ahmed-Fahmy-Dev commented:
@AlinaPalyvoda 
1- text overlay is browser behavior, i have no control over it.
2- fields are disabled in the next step cause the form is stuck in submitting state, if you just hit continue again the inputs will be enabled. anyways i know how to fix it. will do in one of the future commits.
3- i agree, placeholders have to be in a light grey tone, i will figure out how to override mui styles for the textfield label and do it.

Ahmed-Fahmy-Dev committed:
hotfix: checkout parent form
- formModel removed from imports
- formModel removed from props passed to child forms
- activeStep state modified by a callback

Ahmed-Fahmy-Dev committed:
hotfix: child forms
- formModel important locally in each child form
- "required" prop from mui added to the TextFields

Ahmed-Fahmy-Dev committed:
hotfix: checkout form utilities
- spelling mistake corrected in formModel
- * removed from labels since required prop already added
- visa CVV regexp added
- CVV validation edited and expanded

Ahmed-Fahmy-Dev committed:
hotfix: checkout parent form submit handler
- added setSubmitting(false) to fix issue of disabled textFields

Ahmed-Fahmy-Dev committed:
Merge branch 'develop' into feature/cartPage-checkout-form
